### PR TITLE
[Local GC] Move some EE-specific finalize-on-unload logic out of the GC

### PR DESCRIPTION
**Pull request by swgillespie**
`IGCHeap::FinalizeAppDomain` ultimately ends up calling `CFinalize::FinalizeSegForAppDomain`, which finalizes every eligible object on the segment. Today, a little bit of EE introspection is done: if the object we are looking at is "agile and finalizable", it might not get finalized. In particular, if the object being finalized is a thread that is started or some other "agile" non-thread object, it'll get migrated to another app domain.

The GC does not care about these details so this PR moves this logic to the EE as a part of `GCToEEInterface`, since it's an EE-specific detail (`FEATURE_REDHAWK` does not do this check). With this change the GC no longer depends on the Thread class's method table.

### CODE REVIEW

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -36101,43 +36101,21 @@ CFinalize::FinalizeSegForAppDomain (AppDomain *pDomain,
         // if it has the index we are looking for. If the methodtable is null, it can't be from the
         // unloading domain, so skip it.
         if (method_table(obj) == NULL)
+        {
             continue;
+        }
 
-        // eagerly finalize all objects except those that may be agile.
+        // skip objects not in the unloading appdomain
         if (obj->GetAppDomainIndex() != pDomain->GetIndex())
```
Comment (jkotas) [inline src/gc/gc.cpp]: This should be inside the ShouldFinalizeObjectForUnload callback as well. `obj->GetAppDomainIndex()` and `pDomain->GetIndex()` are bad dependencies.

Comment (swgillespie) [inline src/gc/gc.cpp]: will do!

File: src/vm/gcenv.ee.cpp [hunk 2]
```diff
@@ -1349,3 +1349,39 @@ void GCToEEInterface::HandleFatalError(unsigned int exitCode)
 {
     EEPOLICY_HANDLE_FATAL_ERROR(exitCode);
 }
+
+bool GCToEEInterface::ShouldFinalizeObjectForUnload(AppDomain* pDomain, Object* obj)
+{
+    MethodTable *pMT = obj->GetMethodTable();
```
Comment (jkotas) [inline src/vm/gcenv.ee.cpp]: This can just `return true` in CoreCLR - it should be never actually called. We do not have AppDomains in CoreCLR.

Comment (swgillespie) [inline src/vm/gcenv.ee.cpp]: Ok! Are we still using `#ifndef FEATURE_CORECLR` for this, or should I replace the body of this method with `return true`?

Comment (jkotas) [inline src/vm/gcenv.ee.cpp]: You can just replace it with `return true`.

**Comment by swgillespie** [comment]
@Maoni0 PTAL?

File: src/gc/gc.cpp [hunk 3]
```diff
@@ -36101,43 +36101,15 @@ CFinalize::FinalizeSegForAppDomain (AppDomain *pDomain,
         // if it has the index we are looking for. If the methodtable is null, it can't be from the
         // unloading domain, so skip it.
         if (method_table(obj) == NULL)
+        {
             continue;
+        }
 
-        // eagerly finalize all objects except those that may be agile.
-        if (obj->GetAppDomainIndex() != pDomain->GetIndex())
+        // does the EE actually want us to finalize this object? (is it agile?)
```
Comment (Maoni0) [inline src/gc/gc.cpp]: >(is it agile?) [](start = 65, length = 14)

I wouldn't even say (is it agile?) here - it's just decision that EE makes for whatever reason it chooses.

File: src/vm/gcenv.ee.cpp [hunk 4]
```diff
@@ -1349,3 +1349,10 @@ void GCToEEInterface::HandleFatalError(unsigned int exitCode)
 {
     EEPOLICY_HANDLE_FATAL_ERROR(exitCode);
 }
+
+bool GCToEEInterface::ShouldFinalizeObjectForUnload(AppDomain* pDomain, Object* obj)
+{
+    // CoreCLR does not have appdomains, so this code path is dead. Other runtimes may
+    // choose to inspect the object being finalized here.
+    return true;
```
Comment (Maoni0) [inline src/vm/gcenv.ee.cpp]: >  return true; [](start = 2, length = 14)

For changes like this that are very specific to core, you might want to have a uniformed way to annotate them so it'll be easy for you to know to take care of them when we need to port these back to desktop. Perhaps mark with [desktop todo] or something like that.

Comment (swgillespie) [inline src/vm/gcenv.ee.cpp]: will do!

File: src/gc/gc.cpp (hunk 3 above)
Comment (swgillespie) [inline src/gc/gc.cpp]: ok!

**Comment by swgillespie** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**